### PR TITLE
extended availability configuration for dealing with netsplit situations

### PR DESCRIPTION
**Pull request by dergraf**
@larshesel please review. CC: @ioolkos

### CODE REVIEW

File: priv/vmq_server.schema [hunk 1]
```diff
@@ -7,13 +7,29 @@
...
-                                                               {default, off},
-                                                               {datatype, flag}
-                                                              ]}.
+%% @doc Allow new client connections even when a VerneMQ cluster is inconsistent.
+{mapping, "allow_register_during_netsplit", "vmq_server.allow_register_during_netsplit", [
+                                                                                          {default, off},
+                                                                                          {datatype, flag}
+                                                                                         ]}.
+
+%% @doc Allow message publishs even when a VerneMQ cluster is inconsistent.
```
Comment (larshesel) [inline priv/vmq_server.schema:16]: spelling: publishes

File: src/vmq_cluster_com.erl [hunk 2]
```diff
@@ -162,8 +165,8 @@ process(<<"enq", L:32, Bin:L/binary, Rest/binary>>) ->
                           CallerPid ! {Ref, {error, cant_remote_enqueue}}
                   end
           end),
-    process(Rest);
-process(<<>>) -> ok.
+    process(Rest, St);
+process(<<>>, _) -> ok.
```
Comment (larshesel) [inline src/vmq_cluster_com.erl:169]: Moving the RegView from to the process state from the #vmq_msg{} structure is just a refactor and isn't really needed for this change, right? It's fine to refactor it - I just want to know that I didn't miss something.

File: src/vmq_reg.erl [hunk 3]
```diff
@@ -446,92 +454,85 @@ direct_plugin_exports(Mod) when is_atom(Mod) ->
...
+    PublishFun =
+    fun([W|_] = Topic, Payload, Opts) when is_binary(W)
+                                            and is_binary(Payload)
+                                            and is_map(Opts) ->
+            wait_til_ready(),
+            %% allow a plugin developer to override
+            %% - mountpoint
+            %% - dup flag
+            %% - retain flag
+            %% - trade-consistency flag
```
Comment (larshesel) [inline src/vmq_reg.erl:500]: `trade-consistency` no longer exists

Comment (larshesel) [reply src/vmq_reg.erl:500]: Should it be possible to override the new CAP flags?

File: src/vmq_reg.erl [hunk 4]
```diff
@@ -446,92 +454,85 @@ direct_plugin_exports(Mod) when is_atom(Mod) ->
...
+                                         monitor(process, PluginPid),
+                                         plugin_queue_loop(PluginPid, Mod)
+                                 end),
+            QueueOpts = maps:merge(vmq_queue:default_opts(),
+                                   #{clean_session => true,
+                                     is_plugin => true}),
+            {ok, _, _} = register_subscriber(PluginSessionPid, SubscriberId,
+                                             QueueOpts, ?NR_OF_REG_RETRIES),
+            ok
+    end,
```
Comment (larshesel) [inline src/vmq_reg.erl:489]: The RegisterFun seems to be unconditional and not depending on the allow_register_during_netsplit  - while the other functions (publish/subscribe/...) are depending on their respective flags and can return {error, not_ready} depending on those flags. Is this on purpose? I think this was also the old behaviour - but wanted to ask anyway.

**Review by larshesel — COMMENTED** [review]
This was a large one.

Changing this will mean a breaking change for the release it includes as it changes the api to plugins (trade_consistency has been replaced by the cap_settings record). So we'd need to change our lpugins that depend on that (vmq_diversity/webhooks).

Further we need to update the relevant docs.

What else would this impact?

File: priv/vmq_server.schema (hunk 1 above)
Comment (dergraf) [reply priv/vmq_server.schema:16]: No impact on vmq_diversity and vmq_webhooks as those modifiers aren't handled in auth_on_register.  it's not even an specified modifier (outch) https://github.com/erlio/vmq_commons/blob/master/src/auth_on_register_hook.erl

File: src/vmq_cluster_com.erl (hunk 2 above)
Comment (dergraf) [reply src/vmq_cluster_com.erl:169]: yes you're right.

File: src/vmq_reg.erl (hunk 4 above)
Comment (dergraf) [reply src/vmq_reg.erl:489]: It is on purpose. As those  Funs are targeted to be used within plugins I thought the register should always succeed. However for publish and subscribe the plugin might want to treat netsplit situations differently. Now writing this, it sounds like the register fun should be configurable as well. ;) Anyhow, this stuff is currently internal, I think vmq_bridge is the only plugin relying on it (as vmq_systree plugin is gone) . We should improve those functions and make it first-class citizen for plugin developers.

File: src/vmq_reg.erl (hunk 3 above)
Comment (dergraf) [reply src/vmq_reg.erl:500]: why not ;)

**Comment by larshesel** [comment]
Ok, thanks for the clarifications - I don't know if you wanna change stuff (making it possible to override the cap flags) but otherwise it's ok to merge.

**Pull request merged**